src/main/java/eu/dissco/core/digitalspecimenprocessor/web/HandleComponent.java

```diff
@@ -34,12 +37,33 @@ public class HandleComponent {
   private static final String UNEXPECTED_MSG = "Unexpected response from handle API";
   private static final String UNEXPECTED_LOG = "Unexpected response from Handle API. Missing id and/or primarySpecimenObjectId. Response: {}";
 
-  public Map<String, String> postHandle(List<JsonNode> requestBody)
+  public Map<String, String> postHandle(List<JsonNode> request)
       throws PidAuthenticationException, PidCreationException {
+    var requestBody = BodyInserters.fromValue(request);
+    var response = sendRequest(HttpMethod.PATCH, requestBody, "upsert");
+    var responseJsonNode = getFutureResponse(response);
+    return getHandleName(responseJsonNode);
+  }
+
+  public void rollbackHandleCreation(JsonNode request) throws PidCreationException, PidAuthenticationException {
+    var requestBody = BodyInserters.fromValue(request);
+    var response = sendRequest(HttpMethod.DELETE, requestBody, "rollback");
+    getFutureResponse(response);
+  }
+
+  public void rollbackHandleUpdate(List<JsonNode> request) throws PidCreationException, PidAuthenticationException {
+    var requestBody = BodyInserters.fromValue(request);
+    var response = sendRequest(HttpMethod.DELETE, requestBody, "rollback/update");
+    getFutureResponse(response);
+  }
+
+  private <T> Mono<JsonNode> sendRequest(HttpMethod httpMethod, BodyInserter<T, ReactiveHttpOutputMessage> requestBody, String endpoint) throws PidAuthenticationException {
```

Comment:
You are going over line lengths (might need to configure Sonar to trigger on this). Could you reformat?